File: Lista de Exercícios/Java/Empresa.java
import java.util.ArrayList;
import java.util.List;

// Questão 8 e 9
class Empregado {
    private String nome;
    private String cargo;
    private double salario;

    public Empregado(String nome, String cargo, double salario) {
        this.nome = nome;
        this.cargo = cargo;
        this.salario = salario;
    }

    @Override
    public String toString() {
        return String.format("%s - %s (Salário: R$%.2f)", nome, cargo, salario);
    }
}

class Empresa {
    private String nome;
    private List<Empregado> empregados;

    public Empresa(String nome) {
        this.nome = nome;
        this.empregados = new ArrayList<>();
    }

    public void adicionarEmpregado(Empregado empregado) {
        empregados.add(empregado);
    }

    public void listarEmpregados() {
        System.out.println("Empregados da " + nome + ":");
        for (Empregado empregado : empregados) {
            System.out.println(empregado);
        }
    }
}

interface Imprimivel {
    void imprimir();
}

class Relatorio implements Imprimivel {
    private String titulo;

    public Relatorio(String titulo) {
        this.titulo = titulo;
    }

    @Override
    public void imprimir() {
        System.out.println("Imprimindo relatório: " + titulo);
        System.out.println("Conteúdo do relatório...");
    }
}

class Contrato implements Imprimivel {
    private String numero;

    public Contrato(String numero) {
        this.numero = numero;
    }

    @Override
    public void imprimir() {
        System.out.println("Imprimindo contrato número: " + numero);
        System.out.println("Termos e condições do contrato...");
    }
}

public class Empresa {
    public static void imprimirDocumento(Imprimivel doc) {
        doc.imprimir();
    }

    public static void main(String[] args) {
        Empresa empresa = new Empresa("Minha Empresa Ltda.");

        Empregado emp1 = new Empregado("João Silva", "Desenvolvedor", 5000);
        Empregado emp2 = new Empregado("Maria Santos", "Gerente de Projetos", 7000);
        Empregado emp3 = new Empregado("Carlos Oliveira", "Analista de Dados", 4500);

        empresa.adicionarEmpregado(emp1);
        empresa.adicionarEmpregado(emp2);
        empresa.adicionarEmpregado(emp3);

        empresa.listarEmpregados();

        System.out.println("\nTestando a interface Imprimivel:");
        Relatorio relatorio = new Relatorio("Relatório Anual");
        Contrato contrato = new Contrato("C-12345");

        imprimirDocumento(relatorio);
        System.out.println();
        imprimirDocumento(contrato);
    }
}

File: Lista de Exercícios/Golang/Empresa.go
// Questão 8 e 9
package main

import (
	"fmt"
)

// Empregado representa um empregado da empresa
type Empregado struct {
	Nome    string
	Cargo   string
	Salario float64
}

// String retorna uma representação em string do empregado
func (e *Empregado) String() string {
	return fmt.Sprintf("%s - %s (Salário: R$%.2f)", e.Nome, e.Cargo, e.Salario)
}

// Empresa representa uma empresa com seus empregados
type Empresa struct {
	Nome       string
	Empregados []*Empregado
}

// AdicionarEmpregado adiciona um empregado à empresa
func (e *Empresa) AdicionarEmpregado(empregado *Empregado) {
	e.Empregados = append(e.Empregados, empregado)
}

// ListarEmpregados lista todos os empregados da empresa
func (e *Empresa) ListarEmpregados() {
	fmt.Printf("Empregados da %s:\n", e.Nome)
	for _, empregado := range e.Empregados {
		fmt.Println(empregado)
	}
}

// Imprimivel é uma interface para objetos que podem ser impressos
type Imprimivel interface {
	Imprimir()
}

// Relatorio representa um relatório imprimível
type Relatorio struct {
	Titulo string
}

// Imprimir imprime o relatório
func (r *Relatorio) Imprimir() {
	fmt.Printf("Imprimindo relatório: %s\n", r.Titulo)
	fmt.Println("Conteúdo do relatório...")
}

// Contrato representa um contrato imprimível
type Contrato struct {
	Numero string
}

// Imprimir imprime o contrato
func (c *Contrato) Imprimir() {
	fmt.Printf("Imprimindo contrato número: %s\n", c.Numero)
	fmt.Println("Termos e condições do contrato...")
}

// ImprimirDocumento imprime qualquer documento que implemente a interface Imprimivel
func ImprimirDocumento(doc Imprimivel) {
	doc.Imprimir()
}

func main() {
	empresa := &Empresa{Nome: "Minha Empresa Ltda."}

	emp1 := &Empregado{Nome: "João Silva", Cargo: "Desenvolvedor", Salario: 5000}
	emp2 := &Empregado{Nome: "Maria Santos", Cargo: "Gerente de Projetos", Salario: 7000}
	emp3 := &Empregado{Nome: "Carlos Oliveira", Cargo: "Analista de Dados", Salario: 4500}

	empresa.AdicionarEmpregado(emp1)
	empresa.AdicionarEmpregado(emp2)
	empresa.AdicionarEmpregado(emp3)

	empresa.ListarEmpregados()

	fmt.Println("\nTestando a interface Imprimivel:")
	relatorio := &Relatorio{Titulo: "Relatório Anual"}
	contrato := &Contrato{Numero: "C-12345"}

	ImprimirDocumento(relatorio)
	fmt.Println()
	ImprimirDocumento(contrato)
}
